### PR TITLE
Provide charset and collate options only for dialects supporting them

### PR DESCRIPTION
**Pull request by QuentinFarizon**
Fixes https://github.com/sequelize/umzug/issues/612 

Fixes https://github.com/sequelize/umzug/issues/623

### CODE REVIEW

**Comment by WikiRik** [comment]
You'll need to update the minimal type here to add the dialect property;  https://github.com/sequelize/umzug/blob/main/src/storage/sequelize.ts#L21-L30

File: src/storage/sequelize.ts [hunk 1]
```diff
@@ -16,6 +16,9 @@ export type ModelClass = {
 	findAll(options?: {}): Promise<any[]>;
 	create(options: {}): Promise<void>;
 	destroy(options: {}): Promise<void>;
+	dialect: {
+		name: string;
+	};
```
Comment (WikiRik) [inline src/storage/sequelize.ts:21]: You'll need to add this to SequelizeType I think. Did that give different errors?

Comment (QuentinFarizon) [reply src/storage/sequelize.ts:21]: Sorry, fixed the typo

File: src/storage/sequelize.ts [hunk 2]
```diff
@@ -27,8 +27,13 @@ export type SequelizeType = {
 	isDefined(modelName: string): boolean;
 	model(modelName: string): any;
 	define(modelName: string, columns: {}, options: {}): {};
+	dialect: {
+		name: string;
```
Comment (WikiRik) [inline src/storage/sequelize.ts:31]:
```suggestion
	dialect?: {
		name?: string;
```

Make these optional so the other tests won't complain

File: src/storage/sequelize.ts [hunk 3]
```diff
@@ -142,8 +147,8 @@ export class SequelizeStorage implements UmzugStorage {
 				tableName: this.tableName,
 				schema: this.schema,
 				timestamps: this.timestamps,
-				charset: 'utf8',
-				collate: 'utf8_unicode_ci',
+				charset: DIALECTS_WITH_CHARSET_AND_COLLATE.includes(this.sequelize.dialect.name) ? 'utf8' : null,
+				collate: DIALECTS_WITH_CHARSET_AND_COLLATE.includes(this.sequelize.dialect.name) ? 'utf8_unicode_ci' : null,
```
Comment (WikiRik) [inline src/storage/sequelize.ts:151]:
```suggestion
				charset: DIALECTS_WITH_CHARSET_AND_COLLATE.includes(this.sequelize.dialect!.name!) ? 'utf8' : null,
				collate: DIALECTS_WITH_CHARSET_AND_COLLATE.includes(this.sequelize.dialect!.name!) ? 'utf8_unicode_ci' : null,
```

These are always given in the actual sequelize code, just not in our simple mock type

**Comment by WikiRik** [comment]
We're one step further, now the linting issues commence

File: src/storage/sequelize.ts [hunk 4]
```diff
@@ -142,8 +147,8 @@ export class SequelizeStorage implements UmzugStorage {
 				tableName: this.tableName,
 				schema: this.schema,
 				timestamps: this.timestamps,
-				charset: 'utf8',
-				collate: 'utf8_unicode_ci',
+				charset: DIALECTS_WITH_CHARSET_AND_COLLATE.has(this.sequelize.dialect!.name!) ? 'utf8' : null,
+				collate: DIALECTS_WITH_CHARSET_AND_COLLATE.has(this.sequelize.dialect!.name!) ? 'utf8_unicode_ci' : null,
```
Comment (mmkal) [inline src/storage/sequelize.ts:151]: I agree with the the linter here. If dialect is an optional property we shouldn't use !.

Let's make it a variable and null check it:

```ts
const dialectName = this.sequelize.dialect?.name
const hasCharsetAndCollate = dialectName && DIALECTS_WITH_CHARSET_AND_COLLATE.has(dialectName)

// ...

charset: hasCharsetAndCollate ? 'utf8' : null,
collate: hasCharsetAndCollate ? 'utf8_unicode_ci' : null,
```

**Review by mmkal — CHANGES_REQUESTED** [review]

**Review by mmkal — APPROVED** [review]

**Pull request merged**